src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -75,135 +79,185 @@ makeSem = genFreer True
-           VarT n -> Just n
-           _ -> Nothing
-
-  pure
-    . (, tn, Nothing)
-    .  ForallT [PlainTV r] [memberConstraint]
-    .  foldArrows
-    $  ts
-    ++ [resultType]
--- TODO: Although this should never happen, we obviously need a better error message below.
-genType _       = fail "genSig expects a GADT constructor"
-
--- | Turn all (KindedTV tv StarT) into (PlainTV tv) in the given type
--- This can prevent the need for KindSignatures
-simplifyBndrs :: Maybe Type -> Type -> Type
-simplifyBndrs star = everywhere (mkT $ simplifyBndr star)
-
--- | Turn TvVarBndrs of the form (KindedTV tv StarT) into (PlainTV tv)
--- This can prevent the need for KindSignatures
-simplifyBndr :: Maybe Type -> TyVarBndr -> TyVarBndr
-simplifyBndr (Just star) (KindedTV tv k) | star == k = PlainTV tv
-simplifyBndr _ (KindedTV tv StarT) = PlainTV tv
-simplifyBndr _ bndr = bndr
-
--- | Generates a type signature of the form
--- @x :: Member (Effect e) r => a -> b -> c -> Sem r r@.
-genSig :: Con -> Q Dec
-genSig con = do
-  let
-    getConName (ForallC _ _ c) = getConName c
-    getConName (GadtC [n] _ _) = pure n
-    getConName c = fail $ "failed to get GADT name from " ++ show c
-  conName <- getConName con
-  (t, _, k) <- genType con
-  pure $ SigD (getDeclName conName) $ simplifyBndrs k t
 
+------------------------------------------------------------------------------
 -- | Folds a list of 'Type's into a right-associative arrow 'Type'.
 foldArrows :: [Type] -> Type
-foldArrows = foldr1 (AppT . AppT ArrowT)
+foldArrows = foldr1 $ AppT . AppT ArrowT
+
+------------------------------------------------------------------------------
+-- | Extracts name from type variable (possibly nested in signature and/or
+-- some context), returns 'Nothing' otherwise.
+tvName' :: Type -> Maybe Name
+tvName' = \case
+  ForallT _ _ t -> tvName' t
+  SigT t _      -> tvName' t
+  VarT n        -> Just n
+  ParensT t     -> tvName' t
+  _             -> Nothing
+
+------------------------------------------------------------------------------
+-- | Converts equality constraint with type variable to name and type pair or
+-- leaves predicate as is
+eqPairOrCxt :: Pred -> Either (Name, Type) Pred
+eqPairOrCxt p = case asEqualPred p of
+  Just (VarT n, b) -> Left (n, b)
+  Just (a, VarT n) -> Left (n, a)
+  _                -> Right p
+
+------------------------------------------------------------------------------
+-- | Constructs name from base of input name with function applied
+mapNameBase :: (String -> String) -> Name -> Name
+mapNameBase f = mkName . f . nameBase
+
+------------------------------------------------------------------------------
+-- TODO: separate into some module for general utilities (or find them in
+-- available dependencies)
+
+partitionWith :: (a -> Either b c) -> [a] -> ([b], [c])
+partitionWith f = foldl' choose ([], []) where
+  choose = flip $ either (first . (:)) (second . (:)) . f
 
+splitFromEnd :: Int -> [a] -> ([a], [a])
```

Comment:
Maybe just `bimap reverse reverse . split . reverse`?

Reply:
Yeah, I'm just always cautious about expensive `reverse` - If we're not moving this into separate module (where it may be used on more than e.g `5` element lists), then it's probably more sensible implementation.

Reply:
I think this is worth doing. Slow reverse isn't really a problem for n=5. Especially since you don't ever inspect the `snd` of the result, so the compiler won't actually perform it.

Reply:
Is `splitAtEnd n = swap . join bimap reverse . splitAt n . reverse` fine?

Reply:
Is `splitAtEnd n = swap . join bimap reverse . splitAt n . reverse` fine?